Parse/src/main/java/com/parse/ParseObjectSubclassingController.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (c) 2015-present, Parse, LLC.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+package com.parse;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Modifier;
+import java.util.HashMap;
+import java.util.Map;
+
+/* package */ class ParseObjectSubclassingController {
+  private static class ObjectSubclassInfo {
+    public final Class<? extends ParseObject> type;
+    public final String className;
+    public final Constructor<? extends ParseObject> constructor;
+
+    public ObjectSubclassInfo(Class<? extends ParseObject> clazz) throws NoSuchMethodException {
+      type = clazz;
+      className = getClassName(type);
+      constructor = getConstructor(clazz);
+    }
+
+    private static Constructor<? extends ParseObject> getConstructor(Class<? extends  ParseObject> clazz) throws NoSuchMethodException {
+      Constructor<? extends ParseObject> constructor = clazz.getDeclaredConstructor();
+      int modifiers = constructor.getModifiers();
+      if (Modifier.isPublic(modifiers) || (clazz.getPackage().getName().equals("com.parse") &&
+          !(Modifier.isProtected(modifiers) || Modifier.isPrivate(modifiers)))) {
+        return constructor;
+      }
+      throw new NoSuchMethodException();
+    }
+
+    private static String getClassName(Class<? extends ParseObject> clazz) {
+      ParseClassName info = clazz.getAnnotation(ParseClassName.class);
+      if (info == null) {
+        return null;
+      }
+      return  info.value();
+    }
+  }
+
+  private Object mutex;
+  private Map<String, ObjectSubclassInfo> registeredSubclasses;
+
+  public ParseObjectSubclassingController() {
+    mutex = new Object();
+    registeredSubclasses = new HashMap<>();
+  }
+
+  String getClassName(Class<? extends ParseObject> clazz) {
+    return ObjectSubclassInfo.getClassName(clazz);
+  }
+
+  boolean isTypeValid(String className, Class<? extends ParseObject> clazz) {
+    ObjectSubclassInfo info = null;
+
+    synchronized (mutex) {
+      info = registeredSubclasses.get(className);
+    }
+
+    return info == null
+      ? clazz == ParseObject.class
+      : info.type == clazz;
+  }
+
+  void registerSubclass(Class<? extends ParseObject> clazz) {
+    if (!ParseObject.class.isAssignableFrom(clazz)) {
+      throw new IllegalArgumentException("Cannot register a type that is not a subclass of ParseObject");
+    }
+
+    String className = ObjectSubclassInfo.getClassName(clazz);
```

Comment:
we should be consistent with calling into `ObjectSubclassInfo.getClassName(clazz)` vs `ParseObjectSubclassingController.getClassName(clazz)`